MazerPlatformer/GameObject.cs

```diff
@@ -162,57 +167,63 @@ public Component AddComponent(Component.ComponentType type, object value, string
-        
+        protected Either<IFailure, Unit> DrawCentrePoint(SpriteBatch spriteBatch)
+            => EnsureIf(Diganostics.DrawCentrePoint, () =>  spriteBatch.DrawCircle(Centre, 2, 16, Color.Red, 3f))
+                .IgnoreFailure();
+
         // Draw the max point (lower right point)
-        protected void DrawMaxPoint(SpriteBatch spriteBatch) => DoIf(Diganostics.DrawMaxPoint, ()=> spriteBatch.DrawCircle(MaxPoint, 2, 8, Color.Yellow, 3f));
+        protected Either<IFailure, Unit> DrawMaxPoint(SpriteBatch spriteBatch) 
+            => EnsureIf(Diganostics.DrawMaxPoint, () => spriteBatch.DrawCircle(MaxPoint, 2, 8, Color.Yellow, 3f))
+                .IgnoreFailure();
 
         // Draw the bounding box
-        protected void DrawGameObjectBoundingBox(SpriteBatch spriteBatch) => DoIf(Diganostics.DrawGameObjectBounds, ()=> spriteBatch.DrawRectangle(_boundingBox.ToRectangle(), Color.Lime, 1.5f));
+        protected Either<IFailure, Unit> DrawGameObjectBoundingBox(SpriteBatch spriteBatch) 
+            => EnsureIf(Diganostics.DrawGameObjectBounds, () => spriteBatch.DrawRectangle(_boundingBox.ToRectangle(), Color.Lime, 1.5f))
+                .IgnoreFailure();
 
         // Draw the bounding sphere
-        protected void DrawGameObjectBoundingSphere(SpriteBatch spriteBatch) => DoIf(Diganostics.DrawGameObjectBounds, () => spriteBatch.DrawCircle(_centre, BoundingSphere.Radius, 8, Color.Aqua));
+        protected Either<IFailure, Unit> DrawGameObjectBoundingSphere(SpriteBatch spriteBatch)
+            => EnsureIf(Diganostics.DrawGameObjectBounds, () => spriteBatch.DrawCircle(_centre, BoundingSphere.Radius, 8, Color.Aqua))
+                .IgnoreFailure();
 
         // Draw all the diagnostics together
-        protected void DrawObjectDiagnostics(SpriteBatch spriteBatch)
-        {
-            DrawCentrePoint(spriteBatch);
-            DrawMaxPoint(spriteBatch);
-            DrawGameObjectBoundingBox(spriteBatch);
-            DrawGameObjectBoundingSphere(spriteBatch);
-        }
-
-        // Specific game objects need to initialize
-        public virtual void Draw(SpriteBatch spriteBatch)
+        protected Either<IFailure, Unit> DrawObjectDiagnostics(SpriteBatch spriteBatch) =>
+            DrawCentrePoint(spriteBatch)
+                .Bind(unit => DrawMaxPoint(spriteBatch))
+                .Bind(unit => DrawGameObjectBoundingBox(spriteBatch))
+                .Bind(unit=> DrawGameObjectBoundingSphere(spriteBatch));
+
+        // All game objects can ask to draw some text over it if it wants
+        // dependency on Mazer for game font ok.
+        public virtual Either<IFailure, Unit> Draw(SpriteBatch spriteBatch) =>
+            DoIfReturn(!IsNullOrEmpty(InfoText) && Diganostics.DrawObjectInfoText, () => DrawText(spriteBatch))
+                .Bind(unit => DrawObjectDiagnostics(spriteBatch));
+
+        private Either<IFailure, Unit> DrawText(SpriteBatch spriteBatch) => Ensure(() =>
```

Comment:
This Actually looks quite nice - very similar to normal function expect for the function header which now has `=> Ensure() =>` in it